source/common/network/connection_impl.cc

```diff
@@ -438,6 +445,27 @@ void ConnectionImpl::onWriteReady() {
     ASSERT(0 == rc);
     UNREFERENCED_PARAMETER(rc);
 
+#ifdef __APPLE__
+    // When a connect() is performed on a non-blocking socket, it'll return -1/EINPROGRESS
+    // immediately and when select()'ed for writability it'll succeed or fail but when it fails the
+    // reason is hidden. `getsockopt(,,SO_ERROR,,)` solves this issue on some systems but in some
+    // cases (FreeBSD, OS X) getsockopt returns 0 even on `EAGAIN`. The following trick issues a
+    // `getpeername()` which will return -1/ENOTCONN if the socket is not connected and read(,,1)
+    // will produce the correct errno.
+    socklen_t slen = sizeof(struct sockaddr_in);
+    struct sockaddr sain;
+    if (error == 0 && getpeername(fd_, &sain, &slen) == -1 && errno == ENOTCONN) {
+      int rc = read_buffer_->read(fd_, 1);
+      ASSERT(-1 == rc);
+      UNREFERENCED_PARAMETER(rc);
+
+      error = errno;
+      if (errno == EAGAIN) {
+        return;
+      }
+    }
+#endif
```

Comment:
Ooops, sorry! `kq_errno` is indeed just NGINX's field name for that.

What you want is to check `flags` (if it's `EV_EOF`) and `fflags` (for `errno`) fields from `struct kevent`, see `kqueue(2)`.